blog/_posts/2016-10-06-julia-0.5-highlights.md

```diff
@@ -0,0 +1,726 @@
+    double_it_vec(v) = 2v             # vectorized doubling of input
+    double_it_map(v) = map(x->2x, v)  # map a lambda over input
+
+First, a timing comparison in Julia 0.4:
+
+    julia> VERSION
+    v"0.4.7"
+
+    julia> mean([@elapsed(double_it_vec(v)) for _=1:100])
+    0.024444888209999998
+
+    julia> mean([@elapsed(double_it_map(v)) for _=1:100])
+    0.5515606454499999
+
+The functional version defined in terms of `map` is 22 times slower on 0.4 than the vectorized form using specialized generated code for maximal speed.
+Second, the same comparison in Julia 0.5:
+
+    julia> VERSION
+    v"0.5.0"
+
+    julia> mean([@elapsed(double_it_vec(v)) for _=1:100])
+    0.024549842180000003
+
+    julia> mean([@elapsed(double_it_map(v)) for _=1:100])
+    0.023871925960000002
+
+The version defined in terms of `map` is as fast as the vectorized method in 0.5.
+In this case, writing `2v` happens to be more convenient than writing `map(x->2x, v)` – and, of course, you can keep using vectorized forms when they're convenient – but there are many cases where functional constructs are more general, clearer, and more convenient.
+Now they're also fast.
+
+### Ambiguous methods
+
+One design decision that any multiple dispatch language must make is how to handle dispatch ambiguities: cases where none of the methods applicable to a given set of arguments is more specific than the rest.
+Suppose, for example, that a generic function, `f`, has the following methods:
+
+    f(a::Int, b::Real) = 1
+    f(a::Real, b::Int) = 2
+
+In Julia 0.4 and earlier, the second method definition causes an ambiguity warning:
+
+    WARNING: New definition
+        f(Real, Int64) at none:1
+    is ambiguous with:
+        f(Int64, Real) at none:1.
+    To fix, define
+        f(Int64, Int64)
+    before the new definition.
+
+This warning is clear and gets right to the point: the case `f(a,b)` where `a` and `b` are of type `Int` (aka `Int64` on 64-bit systems) is ambiguous.
+Evaluating `f(3,4)` calls the first method of `f` – but this behavior is undefined.
+Giving a warning whenever methods *could* be ambiguous is a fairly conservative choice: it urges people to define a method covering the ambiguous intersection before even defining the methods that overlap.
+When we decided to give warnings for potentially ambiguous methods, we hoped that people would avoid ambiguities and all would be well in the world.
+
+Warning about method ambiguities turns out to be both too strict and too lenient.
+It's far too easy for ambiguities to arise when shared generic functions serve as extension points across unrelated packages.
+When many packages extend the same generic functions, it's *very* common for the methods added to have some ambiguous overlap.
+This happens even when each package has no ambiguities on its own.
+Worse still, slight changes to one package can introduce ambiguities elsewhere, resulting in the least fun game of whack-a-mole ever.
```

Comment:
"whack-a-mole" seems to be a rather local-to-the-US thing. (I mean, I personally don't really understand the reference.)